WalletWasabi/Blockchain/TransactionBuilding/BnB/CheapestSelectionStrategy.cs

```diff
@@ -81,29 +86,61 @@ public EvaluationResult Evaluate(long[] selection, int depth, long sum)
 	{
 		long totalCost = sum + _currentInputCosts;
 
-		if (totalCost > _bestTargetSoFar)
+		if (SuggestionType is SuggestionType.More)
 		{
-			// Our solution is already better than what we might get here.
-			return EvaluationResult.SkipBranch;
+			if (totalCost > _bestTargetSoFar)
+			{
+				// Our solution is already better than what we might get here.
+				return EvaluationResult.SkipBranch;
+			}
+			else if (sum >= Target)
+			{
+				if (_bestTargetSoFar > totalCost)
+				{
+					_bestSelectionSoFar = selection[0..depth];
+					_bestTargetSoFar = totalCost;
+				}
+
+				// Even if a match occurred we cannot be sure that there isn't
+				// a better selection thanks to input costs.
+				return EvaluationResult.SkipBranch;
+			}
+			else if (depth == selection.Length)
+			{
+				// Leaf reached, no match
+				return EvaluationResult.SkipBranch;
+			}
+
+			return EvaluationResult.Continue;
 		}
-		else if (sum >= Target)
+		else if (SuggestionType is SuggestionType.Less)
 		{
-			if (_bestTargetSoFar > totalCost)
+			if (totalCost > Target)
```

Comment:
I would kind of expect `_bestTargetSoFar` here. Can you explain it?

Reply:
Well, because what we are trying to find is a coin combo which is lesser than the `target` but more than the `_bestTargetSoFar`.

Reply:
if you mean that we should change the order of the checks, i'm fine with it.